doc/reference/index.rst

```diff
@@ -23,3 +23,214 @@ API Reference
    usb/index.rst
    usermode/index.rst
    runtime_conf/index.rst
+
+Terminology
+***********
+
+The following terms may be used as shorthand API tags to indicate the
+allowed calling context (thread, ISR, pre-kernel), the effect of a call
+on the current thread state, and other behavioral characteristics.
+
+* :ref:`api_term_reschedule` if executing the function reaches a
+  reschedule point
+* :ref:`api_term_sleep` if executing the function can cause the invoking
+  thread to sleep
+* :ref:`api_term_no-wait` if a parameter to the function can prevent the
+  invoking thread from trying to sleep
+* :ref:`api_term_isr-ok` if the function can always be safely called from non-thread
+  (interrupt) context even if it may return an error in that case
+* :ref:`api_term_pre-kernel-ok` if the function can be safely called
+  before the kernel has been fully initialized, even if it may return an
+  error in that case
+* :ref:`api_term_async` if the function may return before the operation
+  it initializes is complete (i.e. function return and operation
+  completion are asynchronous)
+* :ref:`api_term_supervisor` if the calling thread must have supervisor
+  privileges to execute the function
+
+Details on the behavioral impact of each attribute are in the following
+sections.  Attributes for some common kernel API are below:
+
++-------------------------------+------------+-------+---------+--------+-------+-----------+
```

Comment:
While this works fine as is, such tables can be hard to edit.  If you think this will need updating, easier would be to use a list-table:
```
.. list-table::
   :header-rows: 1

   * - name
     - reschedule
     - sleep
     - no-wait
     - isr-ok
     - async
     - supervisor
   * - :c:func:`k_busy_wait`
     - no
     - no*
     - no*
     - yes*
     - no
     - no
...
```
     
```